wolfcrypt/src/port/arm/thumb2-poly1305-asm_c.c

```diff
@@ -0,0 +1,421 @@
+        /* DONE */
+        "LDM	sp, {r4, r5, r6}\n\t"
+#endif /* WOLFSSL_SP_NO_UMAAL */
+        /* r12 will be zero because r is masked. */
+        /* Load length */
+        "LDR	%[len], [sp, #20]\n\t"
+        /* Reduce mod 2^130 - 5 */
+        "BIC	%[notLast], r8, #0x3\n\t"
+        "AND	r8, r8, #0x3\n\t"
+        "ADDS	r4, r4, %[notLast]\n\t"
+        "LSR	%[notLast], %[notLast], #2\n\t"
+        "ADCS	r5, r5, r9\n\t"
+        "ORR	%[notLast], %[notLast], r9, LSL #30\n\t"
+        "ADCS	r6, r6, r10\n\t"
+        "LSR	r9, r9, #2\n\t"
+        "ADCS	r7, r7, r11\n\t"
+        "ORR	r9, r9, r10, LSL #30\n\t"
+        "ADC	r8, r8, r12\n\t"
+        "LSR	r10, r10, #2\n\t"
+        "ADDS	r4, r4, %[notLast]\n\t"
+        "ORR	r10, r10, r11, LSL #30\n\t"
+        "ADCS	r5, r5, r9\n\t"
+        "LSR	r11, r11, #2\n\t"
+        "ADCS	r6, r6, r10\n\t"
+        "ADCS	r7, r7, r11\n\t"
+        "ADC	r8, r8, r12\n\t"
+        /* Sub 16 from length. */
+        "SUBS	%[len], %[len], #0x10\n\t"
+        /* Store length. */
+        "STR	%[len], [sp, #20]\n\t"
+        /* Loop again if more message to do. */
+#if defined(__GNUC__)
+        "BGT	L_poly1305_thumb2_16_loop_%=\n\t"
+#elif defined(__IAR_SYSTEMS_ICC__) && (__VER__ < 9000000)
+        "BGT.N	L_poly1305_thumb2_16_loop\n\t"
+#else
+        "BGT.N	L_poly1305_thumb2_16_loop_%=\n\t"
+#endif
+        "STM	lr, {r4, r5, r6, r7, r8}\n\t"
+        "\n"
+#if defined(__IAR_SYSTEMS_ICC__) && (__VER__ < 9000000)
+    "L_poly1305_thumb2_16_done:\n\t"
+#else
+    "L_poly1305_thumb2_16_done_%=:\n\t"
+#endif
+        "ADD	sp, sp, #0x1c\n\t"
+        : [ctx] "+r" (ctx), [m] "+r" (m), [len] "+r" (len), [notLast] "+r" (notLast)
+        :
+        : "memory", "r4", "r5", "r6", "r7", "r8", "r9", "r10", "r11", "r12", "lr", "cc"
+    );
+}
+
+XALIGNED(16) static const uint32_t L_poly1305_thumb2_rt[] = {
+    0x0fffffff, 0x0ffffffc, 0x0ffffffc, 0x0ffffffc,
+};
+
+#ifndef WOLFSSL_NO_VAR_ASSIGN_REG
+void poly1305_set_key(Poly1305* ctx_p, const byte* key_p)
+#else
+void poly1305_set_key(Poly1305* ctx, const byte* key)
+#endif /* !WOLFSSL_NO_VAR_ASSIGN_REG */
+{
+#ifndef WOLFSSL_NO_VAR_ASSIGN_REG
+    register Poly1305* ctx __asm__ ("r0") = (Poly1305*)ctx_p;
+    register const byte* key __asm__ ("r1") = (const byte*)key_p;
+    register uint32_t* L_poly1305_thumb2_rt_c __asm__ ("r2") = (uint32_t*)&L_poly1305_thumb2_rt;
+#endif /* !WOLFSSL_NO_VAR_ASSIGN_REG */
+
+    __asm__ __volatile__ (
+        /* Load and cache padding. */
+        "LDR	r2, [%[key], #16]\n\t"
+        "LDR	r3, [%[key], #20]\n\t"
+        "LDR	r4, [%[key], #24]\n\t"
+        "LDR	r5, [%[key], #28]\n\t"
+        "ADD	r10, %[ctx], #0x24\n\t"
+        "STM	r10, {r2, r3, r4, r5}\n\t"
+        /* Load, mask and store r. */
+        "MOV	r10, %[L_poly1305_thumb2_rt]\n\t"
+        "LDR	r2, [%[key]]\n\t"
+        "LDR	r3, [%[key], #4]\n\t"
+        "LDR	r4, [%[key], #8]\n\t"
+        "LDR	r5, [%[key], #12]\n\t"
+        "LDM	r10, {r6, r7, r8, r9}\n\t"
```

Comment:
Getting a fault here...
![Screenshot 2024-09-10 at 4 27 18 PM](https://github.com/user-attachments/assets/ff68b90b-411e-4bfa-8728-4e025404d89b)

Reply:
This is a Cortex M7 (STM32H753ZI). This was with debug build.

```arm-none-eabi-gcc "../Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/wolfcrypt/src/port/arm/thumb2-poly1305-asm_c.c" -mcpu=cortex-m7 -std=gnu11 -g3 -DUSE_HAL_DRIVER -DHAVE_PKCS11_STATIC -DSTM32H753xx -DDEBUG -c -I../Middlewares/Third_Party/FreeRTOS/Source/include -I"/Users/davidgarske/Projects/WolfSSL/STM/STM32H7/STM32H753/Middlewares/Third_Party/wolfPKCS11" -I../Middlewares/Third_Party/FreeRTOS/Source/portable/GCC/ARM_CM4F -I../Drivers/CMSIS/Include -I../Core/Inc -I../Drivers/STM32H7xx_HAL_Driver/Inc/Legacy -I../Drivers/CMSIS/Device/ST/STM32H7xx/Include -I../Middlewares/Third_Party/FreeRTOS/Source/CMSIS_RTOS_V2 -I../Drivers/STM32H7xx_HAL_Driver/Inc -I../Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl -I../wolfSSL/. -I../wolfSSL -I../Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/ -I../Middlewares/Third_Party/wolfSSL_wolfSSH_wolfSSH/wolfssh/ -I../Middlewares/Third_Party/wolfSSL_wolfMQTT_wolfMQTT/wolfmqtt/ -I../wolfTPM -I../Middlewares/Third_Party/wolfSSL_wolfTPM_wolfTPM/wolftpm/ -O0 -ffunction-sections -fdata-sections -Wall -fomit-frame-pointer -fstack-usage -fcyclomatic-complexity -MMD -MP -MF"Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/wolfcrypt/src/port/arm/thumb2-poly1305-asm_c.d" -MT"Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/wolfcrypt/src/port/arm/thumb2-poly1305-asm_c.o" --specs=nano.specs -mfpu=fpv5-d16 -mfloat-abi=hard -mthumb -o "Middlewares/Third_Party/wolfSSL_wolfSSL_wolfSSL/wolfssl/wolfcrypt/src/port/arm/thumb2-poly1305-asm_c.o"```

Reply:
Fixed.